File: Data Structure/Union find/Medium/LET_no947_Most Stones Removed with Same Row or Column.py
class UnionFind:
    def __init__(self, points):
        self.parent = dict()
        for i, j in points:
            self.parent[i], self.parent[~j] = i, ~j
        
        self.size = len(self.parent)
    
    def find(self, x):
        root = x
        while root != self.parent[root]:
            root = self.parent[root]
        
        while x != self.parent[x]:
            temp = self.parent[x]
            self.parent[x] = root
            x = temp
        
        return root
    
    def union(self, x, y):
        px, py = self.find(x), self.find(y)
        if px != py:
            self.parent[px] = py
            self.size -= 1

class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        uf = UnionFind(stones)
        for i, j in stones:
            uf.union(i, ~j)
        
        return len(stones) - uf.size

File: Data Structure/Union find/Medium/LET_no947_Most Stones Removed with Same Row or Column.cpp
class UnionFind {
private:
    unordered_map<int, int> parent;
    int groups;
public:
    UnionFind(vector<vector<int>>& points) {
        for (int i = 0; i < points.size(); i++) {
            int x = points[i][0], y = points[i][1];
            parent[x] = x;
            parent[~y] = ~y;
        }
        groups = parent.size();
    }
    
    int find(int x) {
        int root = x;
        while (root != parent[root])
            root = parent[root];
        
        while (x != parent[x]) {
            int temp = parent[x];
            parent[x] = root;
            x = temp;
        }
        
        return root;
    }
    
    void Union(int x, int y) {
        int px = find(x), py = find(y);
        if (px != py) {
            groups--;
            parent[px] = py;
        }
    }
    
    int getGroupSize(void) { return groups; }
};

class Solution {
public:
    int removeStones(vector<vector<int>>& stones) {
        UnionFind uf = UnionFind(stones);
        for (int i = 0; i < stones.size(); i++) {
            int x = stones[i][0], y = stones[i][1];
            uf.Union(x, ~y);
        }
        
        return stones.size() - uf.getGroupSize();
    }
};